### PR TITLE
Period create/update days

### LINKED ISSUE #2396: Create/Update Period muss anzahl Tage anpassen
Create -> Tage erstellen
Update -> anz. Tage anpassen

### PR DESCRIPTION
**Pull request by pmattmann**
closes #2396

- [x] Create Camp with Period -> create days
- [x] Create Period -> create days
- [x] Update Period -> delete / create days  &  update SchedulEntries
- [x] Validation for Period start/end  with respect to existing ScheduleEntries

Missing:
- [ ] UnitTest for Validation

### CODE REVIEW

File: api/src/DataPersister/PeriodDataPersister.php [hunk 1]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
     }
 
+    public function beforeCreate($data): BaseEntity {
+        static::managePeriodDays($data);
+
+        return $data;
+    }
+
+    public function beforeUpdate($data): BaseEntity {
+        $orig = $this->em->getUnitOfWork()->getOriginalEntityData($data);
```
Comment (pmattmann) [inline api/src/DataPersister/PeriodDataPersister.php:32]: Orig-Daten von Period bestimmen - ist das der korrekte Weg? 
Oder gibt es da eine bessere Alternative?

File: api/src/Entity/Period.php [hunk 2]
```diff
@@ -124,6 +124,15 @@ class Period extends BaseEntity implements BelongsToCampInterface {
     #[Groups(['read', 'write'])]
     public ?DateTimeInterface $start = null;
 
+    /**
+     * If Period-Start date is changing, moveScheduleEntries defines what happens with ScheduleEntries.
+     * true: ScheduleEntries will be moved together with Period (periodOffste stays the same).
+     * false: The start date of each ScheduleEntry remains the same (periodOffset changes).
```
Comment (carlobeltrame) [inline api/src/Entity/Period.php]:
```suggestion
     * If the start date of the period is changing, moveScheduleEntries defines what happens with the schedule
     * entries in the period.
     * true: The schedule entries will be moved together with the period (periodOffset stays the same).
     * false: The start date of each schedule entry remains the same (periodOffset changes).
```

File: api/src/Entity/Period.php [hunk 3]
```diff
@@ -124,6 +124,15 @@ class Period extends BaseEntity implements BelongsToCampInterface {
...
+     */
+    #[ApiProperty(example: true)]
+    #[Groups(['write'])]
+    public bool $moveScheduleEntries = false;
```
Comment (carlobeltrame) [inline api/src/Entity/Period.php]: Ich wäre dafür, das standardmässig auf `true` zu setzen, weil es die "safere" Operation ist, die Schedule Entries mitzuverschieben. Dann bekommt man nicht unerwartet Validierungsfehler, sobald die Validierung implementiert ist. Was meinst du?

File: api/src/DataPersister/PeriodDataPersister.php [hunk 4]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
+        return $data;
+    }
+
     public function beforeRemove($data): ?BaseEntity {
         $this->validator->validate($data, ['groups' => ['delete', 'Period:delete']]);
 
         return null;
     }
+
+    public static function managePeriodDays(Period $period, array $orig = null) {
```
Comment (carlobeltrame) [inline api/src/DataPersister/PeriodDataPersister.php]: Statt einzucodieren dass `$orig` ein `array` sein muss, wie fändest du es wenn wir stattdessen nur `DateTimeInterface $origStart` als Argument verlangen würden?

File: api/src/DataPersister/PeriodDataPersister.php [hunk 5]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
+        if (null != $orig) {
+            $minDelta = $orig['start']->getTimestamp() - $period->start->getTimestamp();
+            $minDelta = floor($minDelta / 60);
+        }
+
+        // Add Days
+        $i = count($days);
+        while ($i < $length) {
+            $day = new Day();
+            $day->dayOffset = $i++;
```
Comment (carlobeltrame) [inline api/src/DataPersister/PeriodDataPersister.php:53]: Wenn ich das Startdatum einer Periode vorverschiebe, und das Enddatum am gleichen Datum belasse, werden die Days inklusive DayResponsibles vorverschoben, und am Ende des Lagers werden neue, leere Days eingefügt. Könnte etwas unintuitiv für API-Clients sein die sich nur genau auf diese Operation fokussieren. Aber ich glaube ist schon gut so, wenn man das big picture ansieht. Wäre ja irgendwie auch komisch wenn wir Spezial-Logik einführen würden, die nur dann greift wenn die alle vormaligen Tage noch in der neuen Datums-Range drin sind.

File: api/tests/Api/Periods/UpdatePeriodTest.php [hunk 6]
```diff
@@ -233,4 +235,84 @@ public function testPatchPeriodReturnsProperDatesInTimezoneBehindUTC() {
...
+            ], 'headers' => ['Content-Type' => 'application/merge-patch+json']])
+        ;
+        $this->assertResponseStatusCodeSame(200);
+
+        /** @var Period $period */
+        $period = $this->getEntityManager()->getRepository(Period::class)->find($period->getId());
+        $this->assertCount(2, $period->days);
+    }
+
+    public function testPatchPeriodMoveScheduleEntrys() {
```
Comment (carlobeltrame) [inline api/tests/Api/Periods/UpdatePeriodTest.php]:
```suggestion
    public function testPatchPeriodMovesScheduleEntries() {
```

File: api/tests/Api/Periods/UpdatePeriodTest.php [hunk 7]
```diff
@@ -233,4 +235,84 @@ public function testPatchPeriodReturnsProperDatesInTimezoneBehindUTC() {
...
+        /** @var Period $period */
+        $period = $this->getEntityManager()->getRepository(Period::class)->find($period->getId());
+        $this->assertCount(1, $period->scheduleEntries);
+
+        /** @var ScheduleEntry $scheduleEntry */
+        $scheduleEntry = $period->scheduleEntries[0];
+        $this->assertEquals(540, $scheduleEntry->periodOffset);
+    }
+
+    public function testPatchPeriodDoNotMoveScheduleEntrys() {
```
Comment (carlobeltrame) [inline api/tests/Api/Periods/UpdatePeriodTest.php]:
```suggestion
    public function testPatchPeriodDoNotMoveScheduleEntries() {
```

File: api/src/DataPersister/PeriodDataPersister.php [hunk 8]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
+            }
+        }
+
+        // Remove Days
+        $i = count($days);
+        while ($i > $length) {
+            $day = $days[--$i];
+            $period->removeDay($day);
+        }
+    }
```
Comment (carlobeltrame) [inline api/src/DataPersister/PeriodDataPersister.php]: Die API Tests gefallen mir schon ziemlich gut, aber trotzdem wäre ich dafür, diese Methode und `Period#getPeriodLength` noch zu unit-testen. Da könnte man dann etwas einfacher auch Edge Cases testen, z.B. wie reagiert `manageDays` wenn `getPeriodLength` `null` zurückgibt?

File: api/src/DataPersister/PeriodDataPersister.php (hunk 1 above)
Comment (carlobeltrame) [reply api/src/DataPersister/PeriodDataPersister.php:32]: Alternativ könntest du dir per DI `public RequestStack $requestStack` injizieren lassen und dann mit `$this->requestStack->getCurrentRequest()->attributes->get('previous_data')` die vorherigen Daten rausholen. Das ist aber vor allem nützlich wenn man auch die vormaligen Relations braucht. Und ausserdem würde das dann das vormalige Camp oder die vormalige Periode geben, je nach dem ob ich gerade auf `/camps/1a2b3c4d` oder `/periods/1a2b3c4d` gepatcht habe (erlauben wir aktuell nicht, aber könnte ja in Zukunft mal sein).

Darum gefällts mir besser so wie dus jetzt gelöst hast.

File: api/src/DataPersister/PeriodDataPersister.php [hunk 9]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
+        $days = $period->getDays();
+
+        $minDelta = 0;
+        if (null != $orig) {
+            $minDelta = $orig['start']->getTimestamp() - $period->start->getTimestamp();
+            $minDelta = floor($minDelta / 60);
+        }
+
+        // Add Days
+        $i = count($days);
```
Comment (carlobeltrame) [inline api/src/DataPersister/PeriodDataPersister.php:50]: Wäre es schlimm wenn wir mal aus Versehen in der DB inkonsistente Daten haben und `count($days)` nicht gleich ist wie der grösste `dayOffset` aller Days dieser Periode?
Ich frage mich, ob es hier angebracht wäre, so einen Zustand zu "reparieren", oder ob das besser anderswo gemacht würde, wenn überhaupt.

**Review by carlobeltrame — CHANGES_REQUESTED** [review]

File: api/src/Entity/Period.php (hunk 3 above)
Comment (pmattmann) [reply api/src/Entity/Period.php]: Ich habe den Default gleich gewählt, wie er aktuell auf dem GUI ist.
Welches von beidem der Standard ist, ist für mich zweitrangig - Hauptsache konsistent.

File: api/src/DataPersister/PeriodDataPersister.php (hunk 1 above)
Comment (pmattmann) [reply api/src/DataPersister/PeriodDataPersister.php:32]: thx

File: api/src/DataPersister/PeriodDataPersister.php (hunk 5 above)
Comment (usu) [reply api/src/DataPersister/PeriodDataPersister.php:53]: Eigentlich würde es Sinn machen, das Verhalten von `moveScheduleEntries` abhängig zu machen:
Wenn `moveScheduleEntries=true`, dann wäre das jetzige Verhalten erwartet. ScheduleEntries + DayResponsibles werden verschoben und am Ende gibt es neue Days.

Wenn `moveScheduleEntries=false` ist (z.B. ich möchte die Anreise am Tag vor dem Lager noch hinzufügen, alles andere bleibt aber gleich), dann wäre es eigentlich logischer, wenn die DayResponsibles stehen bleiben, und am Anfang ein neuer Tag hinzugefügt wird.

File: api/src/DataPersister/PeriodDataPersister.php (hunk 4 above)
Comment (pmattmann) [reply api/src/DataPersister/PeriodDataPersister.php]: Meine Überlegung war folgende:
Der Aufrufer muss nicht wissen, worauf es ankommt - das ist nicht seinen Aufgabe.
Zudem ist es so einfacher erweiterbar, falls zusätzliche Anforderungen erscheinen.

File: api/tests/Api/Periods/CreatePeriodTest.php [hunk 10]
```diff
@@ -212,6 +212,20 @@ public function testCreatePeriodValidatesStartAfterEnd() {
         ]);
     }
 
+    public function testCraetePeriodCreatesDays() {
```
Comment (usu) [inline api/tests/Api/Periods/CreatePeriodTest.php]:
```suggestion
    public function testCreatePeriodCreatesDays() {
```

File: api/src/DataPersister/PeriodDataPersister.php (hunk 5 above)
Comment (pmattmann) [reply api/src/DataPersister/PeriodDataPersister.php:53]: Guter Punkt!
Spätestens wenn wir Tages-Jobs implementieren werden wir wieder an dieser Frage vorbei kommen.
Ich sehe hier folgende Möglichkeiten:
- Tages-Jobs werden immer mit dem Start-Datum mit geschoben 
   (so ist es aktuell; da an Day-Objekt angehängt; aber evtl. verwirrend)
- Tages-Jobs bleiben immer am gleichen Datum
- Benutzer-Eingabe für ScheduleEntries auch für Jobs verwenden  (mein Favorit)
- Zusätzliche Eingabe von Benutzer verlangen

File: api/src/Entity/Period.php (hunk 3 above)
Comment (pmattmann) [reply api/src/Entity/Period.php]: default von $moveScheduleEntries ist jetzt TRUE

File: api/src/DataPersister/PeriodDataPersister.php (hunk 5 above)
Comment (carlobeltrame) [reply api/src/DataPersister/PeriodDataPersister.php:53]: Dann müssten wir bei `moveScheduleEntries=false` zusätzlich auch validieren, dass keine Tagesverantwortlichen auf Tagen stehen, die gelöscht werden.

File: api/src/DataPersister/PeriodDataPersister.php (hunk 4 above)
Comment (usu) [inline api/src/DataPersister/PeriodDataPersister.php]: Wäre der Funktionsname `updateDaysAndScheduleEntries` treffender? Oder aufbrechen in 2 separate Funktionen `updateDays` und `updateScheduleEntries`?

File: api/src/DataPersister/PeriodDataPersister.php (hunk 5 above)
Comment (usu) [reply api/src/DataPersister/PeriodDataPersister.php:53]: Müssten wir wahrscheinlich sowieso einführen, unabhängig von dieser Diskussion: Ich kann jetzt aktuell eine Periode kürzen, ohne das Startdatum zu ändern. Dann werden am Ende die Days gelöscht, incl. DayResponsibles. 

Oder wir entscheiden uns dafür, dass der Informationsverlust in diesem Fall gering ist und wir erlauben löschen von Tagen, selbst wenn da noch DayReponsibles drauf sind.

Comment (pmattmann) [reply api/src/DataPersister/PeriodDataPersister.php:53]: Das fände ich eher seltsam, wenn man vorher alle Job-Zuteilungen manuell löschen muss, bevor man die Periode verschieben kann.
Da würde ich mich spontan eher für eine Warnung aussprechen.

File: api/src/DataPersister/PeriodDataPersister.php (hunk 4 above)
Comment (pmattmann) [reply api/src/DataPersister/PeriodDataPersister.php]: `updateDaysAndScheduleEntries` finde ich gut.
Trennen wird mühsam, da Tage erstellen vorher und Tage löschen nachher ist.

File: api/src/DataPersister/PeriodDataPersister.php (hunk 5 above)
Comment (usu) [reply api/src/DataPersister/PeriodDataPersister.php:53]: Good for me. Allenfalls müssen wir die Präsentation im Frontend anpassen um bewusster zu kommunizieren, was gerade passiert. Anstatt frei Start und End-datum editieren zu können (aktuelle Implementation) könnten wir die z.B. die folgenden Operationen anbieten:

- Start-Datum anpassen // Default; request wird mit moveScheduleEntires=true abgesetzt
- x Tage am Anfang oder Start hinzufügen // Benutzer kann die Anzahl Tage auswählen; request wird mit moveScheduleEntires=false abgesetzt
- x Tage am Anfang oder am Start löschen // fette Warnung, dass damit möglicherweise Inhalt gelöscht wird; Benutzer kann die Anzahl Tage auswählen; request wird mit moveScheduleEntries=false abgesetzt

Je nach dem, was man erreichen möchte, braucht es dann mehrere Requests. Ist dafür expliziter.

File: api/src/DataPersister/PeriodDataPersister.php (hunk 4 above)
Comment (usu) [reply api/src/DataPersister/PeriodDataPersister.php]: Ok for me. Wieso ist die genaue Reihenfolge relevant? Das Hinzufügen/Löschen von Days kann man doch beides vor Aktualisieren von periodOffsets machen, ohne dabei inkonsistente Daten zu generieren?

File: api/src/Validator/Period/AssertValidPeriodEndValidator.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            if (!$period->moveScheduleEntries) {
+                $orig = $this->em->getUnitOfWork()->getOriginalEntityData($period);
+                if (null != $orig) {
+                    $delta = $orig['start']->getTimestamp() - $period->start->getTimestamp();
+                    $delta = floor($delta / 60);
+                }
+            }
+
+            // get maximal existing ScheduleEntryEnd
+            $scheduleEntryEnds = $period->scheduleEntries->map(fn ($se) => $se->periodOffset + $se->length);
```
Comment (usu) [inline api/src/Validator/Period/AssertValidPeriodEndValidator.php]: Ich hätte persönlich eher mit `DateTime` und `DateInterval` gearbeitet, anstatt manuell zwischen Minuten/Tagen hin- und her zu rechnen. Aber ist wohl Geschmacksache. It get's the job done.

```php
            $delta = $period->start->diff($orig['start'])
            
            $scheduleEntryEnds = $period->scheduleEntries->map(fn ($se) => $se->getEnd());
            $maxScheduleEntryEnd = max($scheduleEntryEnds->toArray())->add($delta);

            if ($maxScheduleEntryEnd > $value) {
```

File: api/src/Validator/Period/AssertValidPeriodEndValidator.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            $maxScheduleEntryEnd = max($scheduleEntryEnds->toArray()) + $delta;
+            $periodEnd = 1440 * $period->getPeriodLength();
+
+            if ($maxScheduleEntryEnd > $periodEnd) {
+                /** @var DateTime $endDate */
+                $endDate = clone $period->end;
```
Comment (usu) [inline api/src/Validator/Period/AssertValidPeriodEndValidator.php]: Sollte das nicht vom start aus rechnen?
```php
$endDate = clone $period->start;
```

File: api/src/Entity/Period.php [hunk 13]
```diff
@@ -116,6 +118,7 @@ class Period extends BaseEntity implements BelongsToCampInterface {
      * @ORM\Column(type="date")
      */
     #[Assert\LessThanOrEqual(propertyPath: 'end')]
+    #[AssertValidPeriodStart()]
```
Comment (usu) [inline api/src/Entity/Period.php]: Finde den Namen sehr generisch und nicht offensichtlich, was hier genau validiert werden soll. Und `$value` wird im Validator aktuell nicht genutzt, d.h. du könntest den Validator im Moment auf irgend ein Property setzten.

Wie wäre es mit:
- AssertValidPeriodStart --> AssertLessThanOrEqualToEarliestScheduleEntryStart
- AssertValidPeriodEnd --> AssertGreaterThanOrEqualToLastScheduleEntryEnd

**Review by usu — CHANGES_REQUESTED** [review]

File: api/src/Validator/Period/AssertValidPeriodEndValidator.php (hunk 12 above)
Comment (pmattmann) [reply api/src/Validator/Period/AssertValidPeriodEndValidator.php]: Habe die Logik umgebaut, so dass diese verständlicher sein sollte.

**Comment by pmattmann** [comment]
Vermutlich ist [PeriodDataPerisister::updateDaysAndScheduleEntries](https://github.com/ecamp/ecamp3/pull/2437/files#diff-767c69164e87f31e4cdf29c976d17c78affb3df872859bc55e058f2933d69370R45) noch nicht perfekt.
Ich würd den PR aber dennoch gerne mergen.

Der Inhalt ändert mit dem Umbau auf ScheduleEntry.StartOffset und EndOffset ohnehin noch einmal.

**Review by usu — APPROVED** [review]
Von mir aus können wir mergen, um #594 nicht im Weg zu stehen.

Für die 2 offenen Punkte gibt's dann noch separate Issues oder Pull Requests?
- Unit test für validators
- Move `DayResponsibles`/`Days` abhängig von `moveScheduleEntries`

File: api/src/Entity/Period.php [hunk 14]
```diff
@@ -249,6 +263,20 @@ public function removeMaterialItem(MaterialItem $materialItem): self {
         return $this;
     }
 
+    /**
+     * Returns the length of this period
```
Comment (carlobeltrame) [inline api/src/Entity/Period.php]:
```suggestion
     * Returns the length of this period in days
```

File: api/src/DataPersister/PeriodDataPersister.php [hunk 15]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
+
+    public static function updateDaysAndScheduleEntries(Period $period, array $orig = null) {
+        $length = $period->getPeriodLength();
+        $days = $period->getDays();
+
+        $minDelta = 0;
+        if (null != $orig) {
+            $minDelta = $orig['start']->getTimestamp() - $period->start->getTimestamp();
+            $minDelta = floor($minDelta / 60);
+        }
```
Comment (carlobeltrame) [inline api/src/DataPersister/PeriodDataPersister.php]: Ich musste lange überlegen bis ich gemerkt habe dass das nicht das "minimale Delta" heissen soll.

```suggestion
        $deltaMinutes = 0;
        if (null != $orig) {
            $deltaMinutes = $orig['start']->getTimestamp() - $period->start->getTimestamp();
            $deltaMinutes = floor($deltaMinutes / 60);
        }
```

File: api/src/DataPersister/PeriodDataPersister.php [hunk 16]
```diff
@@ -19,9 +22,56 @@ public function __construct(
...
+        while ($i < $length) {
+            $day = new Day();
+            $day->dayOffset = $i++;
+            $period->addDay($day);
+        }
+
+        // Move Schedule-Entries
+        if (!$period->moveScheduleEntries) {
+            foreach ($period->scheduleEntries as $scheduleEntry) {
+                $scheduleEntry->periodOffset += $minDelta;
```
Comment (carlobeltrame) [inline api/src/DataPersister/PeriodDataPersister.php]:
```suggestion
                $scheduleEntry->periodOffset += $deltaMinutes;
```

File: api/src/DataPersister/PeriodDataPersister.php (hunk 15 above)
Comment (carlobeltrame) [reply api/src/DataPersister/PeriodDataPersister.php]: Vielleicht diese Berechnung ins `if (!$period->moveScheduleEntries) { ... }` verschieben? Brauchts ja erst dort, und dann wäre es näher an der eigentlichen Benutzung dran.

File: api/src/Validator/Period/AssertGreaterThanOrEqualToLastScheduleEntryEnd.php [hunk 17]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Validator\Period;
+
+use Attribute;
+use Symfony\Component\Validator\Constraint;
+
+#[Attribute]
+class AssertGreaterThanOrEqualToLastScheduleEntryEnd extends Constraint {
+    public string $message = 'Due to existing activities, end-date can not be earlier then {{ endDate }}';
```
Comment (carlobeltrame) [inline api/src/Validator/Period/AssertGreaterThanOrEqualToLastScheduleEntryEnd.php]:
```suggestion
    public string $message = 'Due to existing schedule entries, end-date can not be earlier then {{ endDate }}';
```
"Existing activities" ist auf französisch genauso unklar wie "existing schedule entries": Zumindest im Kurs heisst ein Block "point de cours", nicht "activité". Daher würde ich vom Backend lieber die technisch korrekte englische Nachricht herausgeben, und es dem Frontend durch separate Übersetzungen überlassen, das für Normaluser besser verständlich zu machen.

File: api/src/Validator/Period/AssertLessThanOrEqualToEarliestScheduleEntryStart.php [hunk 18]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Validator\Period;
+
+use Attribute;
+use Symfony\Component\Validator\Constraint;
+
+#[Attribute]
+class AssertLessThanOrEqualToEarliestScheduleEntryStart extends Constraint {
+    public string $message = 'Due to existing activities, start-date can not be later then {{ startDate }}';
```
Comment (carlobeltrame) [inline api/src/Validator/Period/AssertLessThanOrEqualToEarliestScheduleEntryStart.php]:
```suggestion
    public string $message = 'Due to existing schedule entries, start-date can not be later then {{ startDate }}';
```

**Review by carlobeltrame — APPROVED** [review]
